Initial PHP 8.1 support

boesing opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| BC Break      |  no
| New Feature   | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

This provides initial support for PHP 8.1.
This setup is very basic and currently only supports pre-compiled extensions by https://github.com/shivammathur/php-builder.
These extensions are:

```
amqp
apcu
bcmath
bz2
calendar
Core
ctype
curl
date
dom
enchant
exif
FFI
fileinfo
filter
ftp
gd
gettext
gmp
hash
iconv
igbinary
imagick
intl
json
ldap
libxml
mbstring
memcache
memcached
msgpack
mysqli
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_pgsql
pdo_sqlite
pgsql
Phar
posix
pspell
readline
redis
Reflection
session
shmop
SimpleXML
soap
sockets
sodium
SPL
sqlite3
standard
sysvmsg
sysvsem
sysvshm
tidy
tokenizer
xml
xmlreader
xmlwriter
xsl
zip
zlib
```

The `xdebug` extension is also available but disabled by default (the same way as in all other PHP versions).
`pcov` is also available and can be enabled.

If any project needs more than these extensions, they need to install these by themselves by using the `pre-run.sh` script which is being executed on each run.

boesing committed:
docs: fix `dependencies` configuration entry

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: bump swoole extension to `4.7.0`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: normalize `mods-available` synchronization

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add dedicated strategy setups

- Every PHP version has its own setup strategy
- Base system has dedicated `ubuntu` strategy
- Markdownlint has its own strategy

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: introduce setup strategy for PHP 8.1

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: rework `extensions` handling

This will check if an extension is already enabled by the PHP version compiled within this action.
Afterwards, it will check if it is pre-compiled but not yet enabled for the PHP version.
If both checks fail, an extension is marked to be installed.

For PHP 8.1, we do not yet support dynamic installation of extensions as we do for earlier PHP versions. This is due to the fact that PHP 8.1 is not yet provided by the Sury debian repository.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: use glob rather than `ls` to iterate over files in `conf.d`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: `$mod` contains the full qualified path to that file and thus that needs to be addressed

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

weierophinney commented on mods-install/install_swoole.sh:
```diff
@@ -19,13 +19,10 @@ for PHP_VERSION in 7.3 7.4 8.0;do
     make
     make install
+
+    cp /mods-available/swoole.ini /etc/php/${PHP_VERSION}/mods-available/swoole.ini
```
:+1:

weierophinney commented on scripts/extensions.sh:
```diff
@@ -6,23 +6,36 @@
+            echo "Cannot install extension for the current PHP version."
+            echo "Please use \".laminas-ci/pre-run.sh\" to setup specific extensions for PHP $PHP"
+            echo "Additional details can be found on https://stackoverflow.com/q/8141407"
+            echo "The following extensions were not installed: ${EXTENSIONS[@]}"
```
Add `# shellcheck disable=SC2145` above this line - this is an approrpriate place to expand the array within quotes.

boesing replied:
I've used the suggestion from shellcheck (`*` instead of `@`).
That should work as well but will test that tomorrow.

boesing replied:
Works. :)

boesing committed:
bugfix: handle latest changes made by `shivammathur/php-builder`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

weierophinney commented on scripts/extensions.sh:
```diff
@@ -34,64 +47,38 @@ function enable_static_extension {
-    local -a EXTENSIONS=("${@:3}")
-    local TO_RETURN
+# Loop through known statically compiled/installed extensions, and enable them.
+for EXTENSION in ${EXTENSIONS[@]}; do
```
Based on what I know of bash, and a large part of that at this point in time has been via going through shellcheck error messages to see _why_ things are being flagged... quoting `${EXTENSIONS[@]}` here is the right thing to do, as what it does is quote _each entry_ in the array. This ensures that any elements of the array that have whitespace in them are treated as individual items, versus multiple items.

As an example, if  you defined:

```bash
EXTENSIONS=("foo" "foo bar" "bar")
```

then the following:

```bash
for EXTENSION in ${EXTENSIONS[@]}; do
    echo "${EXTENSION}"
done
```

would result in

```txt
foo
foo
bar
bar
```

while

```bash
for EXTENSION in "${EXTENSIONS[@]}";do
    echo "${EXTENSION}"
done
```

would result correctly in

```txt
foo
foo bar
bar
```

I recognize you saw a difference on Mac - I'm not sure why that is, but on the bash shipped with Ubuntu, the above is definitely true.

While we likely won't have those scenarios, this protects us from them.

weierophinney commented on scripts/extensions.sh:
```diff
@@ -34,64 +47,38 @@ function enable_static_extension {
-        enable_static_extension "${PHP}" swoole
+    # Check if extension is already enabled
+    # shellcheck disable=SC2236,SC2143
+    if [ ! -z "$(echo "${ENABLED_EXTENSIONS}" | grep "${EXTENSION}")" ]; then
```
I'd love to figure out a way to do this without needing to suppress the shellcheck warnings, but I tried a bunch of variations, none of which worked as intended.

boesing replied:
I really hate using `echo` to search for stuff but I havent figured out a better way. My bash scripting skills are a bit rusty.

weierophinney replied:
Actually, there IS a way to do this without pipe + grep, since what we're trying to determine is simply whether or not `$ENABLED_EXTENSIONS` contains `$EXTENSION`:

```bash
if [[ "${ENABLED_EXTENSIONS}" =~ "${EXTENSION}" ]]; then
```

boesing replied:
Yah, but sadly, this does not support "words" (and this is not achieved with `grep` either).
So this might lead to problems when searching for exact matches.

Do you have an idea how to solve that?

boesing replied:
Okay, I managed to create a boundary expression.

weierophinney commented on setup/php/8.1/setup.sh:
```diff
@@ -0,0 +1,32 @@
+mkdir -p /etc/php/8.1/{cli,mods-available}
+ln -s $CONFIGURATION_DIRECTORY_PREFIX/php.ini /etc/php/8.1/cli/
+ln -s $CONFIGURATION_DIRECTORY_PREFIX/conf.d /etc/php/8.1/cli/conf.d
+for mod in $(ls /etc/php/8.1/cli/conf.d); do
```
Pay attention to the shellcheck output here - will definitely ensure this doesn't blow up on us!

weierophinney commented on setup/php/8.1/setup.sh:
```diff
@@ -0,0 +1,32 @@
+            # Do not enable these modules as default
+        ;;
+        *)
+            ln -s /etc/php/8.1/mods-available/$modWithoutPriority /etc/php/8.1/cli/conf.d/$mod
```
```suggestion
            ln -s "/etc/php/8.1/mods-available/${modWithoutPriority}" "/etc/php/8.1/cli/conf.d/${mod}"
```

weierophinney reviewed (changes requested):
I'll have time tomorrow to review again, and/or assist - and THANK YOU! This is awesome. :tada:

boesing committed:
bugfix: add missing shebang

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: apply latest shellcheck hints

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: use quoted variables

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove `shellcheck disable`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: disable `SC2005` and add quotes for `basename`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: fix typo

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
/cc @Ocramius
Is there some configuration I can add so that docker wont run all of the `setup/*.sh` commands (which I've added one-by-one to make use of build caches) when *anything* within the `setup/` folder has changed?
I have a huge lack when it comes to docker...

boesing committed:
bugfix: do not disable any modules

Modules may depend on each other and thus, if one is enabled (e.g. `redis`), it will also require `msgpack`, `igbinary` and stuff to be enabled on PHP 8.1

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: avoid using `echo` and `grep` to determine if an extension is already enabled

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
@weierophinney @Ocramius 
Just curious, how can we force rebuild when caches are enabled?
None of our files will change when merging this and with #44, we do cache (whatever? I dont even know what is cachable 😂) stuff?
So if any of the Sury packages do change within the registry and/or shivammathur/php-builder does create new releases, our caches might hit and we won't build a new container, right? Or did I miss-understood on what we are caching with #44?

Maybe adding `workflow_dispatch` (https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/) could prevent caches from being hit so we are able to rebuild?

Well, that wont trigger a new release - but something might work 🤔 
Because closing empty milestones will result in an error. So what would be an appropriate workflow here to build a new container with latest PHP versions?

weierophinney commented:
@boesing So, it turns out @Ocramius was addressing this in #44, which I've just merged with changes.

That said: in order to start caching, we need either (a) a release, or (b) a patch made by somebody as a branch DIRECTLY on the repository. This is due to the fact that GH no longer injects repository secrets into pull_request workflows initiated from a forked repository (i.e., the normal way we do PRs). If the PR is from a branch pushed to the repo already, it can use those secrets ... which in turn allows us to cache layers, as we'll have a logged in user.

So, my suggestion:

- Rebase from current 1.11.x branch
- Push the branch to _this_ repository
- Update the PR to use the branch on this repository as its source, instead of the one on your fork

That will trigger a build, and _should_ allow us to see caching in action; hopefully then the release will be pulling from the cache, which will make it faster.

boesing commented:
Superseded by #48
